dia-batching-server/src/api/coingecko.rs

```diff
@@ -0,0 +1,310 @@
+			("Moonbeam", "GLMR") => Some("moonbeam".to_string()),
+			("Polkadex", "PDEX") => Some("polkadex".to_string()),
+			("Stellar", "XLM") => Some("stellar".to_string()),
+			_ => None,
+		}
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct SimplePing {
+	pub gecko_says: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CoingeckoPrice {
+	pub usd: Decimal,
+	pub usd_market_cap: Option<Decimal>,
+	pub usd_24h_vol: Option<Decimal>,
+	pub usd_24h_change: Option<Decimal>,
+	pub last_updated_at: u64,
+}
+
+/// CoinGecko network client
+pub struct CoingeckoClient {
+	host: String,
+	api_key: String,
+}
+
+impl CoingeckoClient {
+	pub fn new(host: String, api_key: String) -> Self {
+		CoingeckoClient { host, api_key }
+	}
+
+	async fn get<R: DeserializeOwned>(&self, endpoint: &str) -> Result<R, CoingeckoError> {
+		let mut headers = reqwest::header::HeaderMap::new();
+		headers.insert("accept", reqwest::header::HeaderValue::from_static("application/json"));
+
+		// We supply a different header for the demo API
+		let mut api_key = reqwest::header::HeaderValue::from_str(self.api_key.as_str())
+			.map_err(|e| CoingeckoError(format!("Could not set API key header value: {}", e)))?;
+		api_key.set_sensitive(true);
+		if self.host.contains("pro-api") {
+			headers.insert("x-cg-pro-api-key", api_key);
+		} else {
+			headers.insert("x-cg-demo-api-key", api_key);
+		}
+
+		let client = reqwest::Client::builder()
+			.default_headers(headers)
+			.build()
+			.map_err(|e| CoingeckoError(e.to_string()))?;
+
+		let url = reqwest::Url::parse(
+			format!("{host}/{ep}", host = self.host.as_str(), ep = endpoint).as_str(),
+		)
+		.expect("Invalid URL");
+
+		let response = client.get(url).send().await;
+
+		match response {
+			Ok(response) => {
+				if !response.status().is_success() {
+					let result = response.text().await;
+					let err_msg = format!(
+						"CoinGecko API error: {}",
+						result.unwrap_or("Unknown".to_string()).trim()
+					);
+					Err(CoingeckoError(err_msg))
+				} else {
+					let result = response.json().await;
+					result.map_err(|e| {
+						CoingeckoError(format!("Could not decode CoinGecko response: {}", e))
+					})
+				}
+			},
+			Err(e) => Err(CoingeckoError(e.to_string())),
+		}
+	}
+
+	/// Check API server status
+	#[allow(dead_code)]
+	pub async fn ping(&self) -> Result<SimplePing, CoingeckoError> {
+		self.get("/ping").await
+	}
+
+	/// Get the current price of any cryptocurrencies vs USD with full precision
+	pub async fn price<Id: AsRef<str>>(
+		&self,
+		ids: &[Id],
+		include_market_cap: bool,
+		include_24hr_vol: bool,
+		include_24hr_change: bool,
+		include_last_updated_at: bool,
+	) -> Result<HashMap<String, CoingeckoPrice>, CoingeckoError> {
+		let ids = ids.iter().map(AsRef::as_ref).collect::<Vec<_>>();
+		// We always query for USD
+		let vs_currencies = vec!["usd"];
+		let vs_currencies = vs_currencies.iter().map(AsRef::as_ref).collect::<Vec<_>>();
```

Comment:
`vs_currencies` is already a **`Vec<str>`**. You can remove line 188.

Reply:
True, thanks 👍